module/VuFindApi/src/VuFindApi/Controller/ApiController.php

```diff
@@ -40,7 +40,7 @@
  */
 class ApiController extends \VuFind\Controller\AbstractBase
 {
-    use ApiTrait;
+    //use ApiTrait;
```

Comment:
Why comment this out? I think we need this trait, since it sets up event hooks for API security.

Reply:
But the permission checking is done only in SearchApiController (and Search2ApiController and WebApiController), not in ApiController itself... the only thing which is used from ApiTrait is onDispatch method, but do we really need it here? I am not sure, maybe @EreMaijala would know...

The problem found by PHPStan was that it use constant self::STATUS_ERROR, which is defined by ApiInterface, but not in ApiController...

Reply:
So I think we should remove ApiTrait from ApiController, but if onDispatch method is still needed here, we should refactor it somehow...

Reply:
OnDispatch is needed at least for the CORS headers. Otherwise any external Swagger or another in-browser OpenAPI generator would be unable to access the API definitition. The easy, though not particularly pretty, fix would be to move STATUS_OK and STATUS_ERROR to ApiTrait as static variables like so:

```
trait ApiTrait
{
    // define some status constants
    public static $STATUS_OK = 'OK';       // good
    public static $STATUS_ERROR = 'ERROR'; // bad
...
        if (!$auth->isGranted($permission)) {
            return $this->output([], self::$STATUS_ERROR, 403, 'Permission denied');
        }
```

This wouldn't be back-compatible if we get rid of the consts in ApiInterface, but I'd be okay with that.

Reply:
Could we also solve the problem by replacing `self::STATUS_ERROR` with the more explicit `ApiInterface::STATUS_ERROR`? I don't think we really need to reference the constant using self in this instance.

Reply:
Thans @EreMaijala for clarification, and @demiankatz for solution ;) The other solution could be to create another trait and move the onDispatch method there, then use this new trait in ApiController and both in SearchApiController, but your solution is easier